Frameworks/Foundation/NSOperation.mm

```diff
@@ -77,41 +61,88 @@ + (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
     return NO;
 }
 
-/**
- @Status Interoperable
-*/
-+ (id)allocWithZone:(NSZone*)zone {
-    NSOperation* ret = [super allocWithZone:zone];
+- (void)_checkReady {
+    bool newReady = YES;
 
-    ret->priv = new NSOperationPriv();
+    // If cancelled, skip this logic and set _ready to YES. Otherwise check dependencies.
+    if (![self isCancelled]) {
+        _dependenciesLock.lock();
+        int count = [_dependencies count];
 
-    return ret;
+        for (int i = 0; i < count; i++) {
+            id op = [_dependencies objectAtIndex:i];
+
+            if (![op isFinished]) {
+                newReady = NO;
+            }
+        }
+
+        _dependenciesLock.unlock();
+    }
+
+    if (_ready != newReady) {
+        [self willChangeValueForKey:@"isReady"];
+        _ready = newReady;
+        [self didChangeValueForKey:@"isReady"];
+    }
+}
+
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        if (object == self) {
+            _finishLock.lock();
+            _finishCondition.notify_all();
+            _finishLock.unlock();
+        } else {
+            [self _checkReady];
+        }
+    }
+}
+
+- (id)init {
+    if (self = [super init]) {
+        _dependencies.attach([NSMutableArray new]);
+        _ready = YES;
+        [self addObserver:self forKeyPath:@"isFinished" options:0 context:NULL];
+    }
+
+    return self;
 }
 
 /**
  @Status Interoperable
 */
-- (NSOperationQueuePriority)queuePriority {
-    return priv->priority;
+- (BOOL)isReady {
+    return _ready;
 }
 
 /**
  @Status Interoperable
 */
 - (void)addDependency:(id)operation {
-    if (priv->dependencies == nil) {
-        priv->dependencies = [[NSMutableArray alloc] init];
-    }
-    [priv->dependencies addObject:operation];
+    [self willChangeValueForKey:@"dependencies"];
+    _dependenciesLock.lock();
+    [_dependencies addObject:operation];
+    [operation addObserver:self forKeyPath:@"isFinished" options:0 context:NULL];
+    _dependenciesLock.unlock();
+    [self didChangeValueForKey:@"dependencies"];
+    [self _checkReady];
 }
 
 /**
  @Status Interoperable
 */
-- (void)setQueuePriority:(NSOperationQueuePriority)priority {
-    priv->priority = priority;
+- (void)removeDependency:(NSOperation*)operation {
+    [self willChangeValueForKey:@"dependencies"];
+    _dependenciesLock.lock();
+    [operation removeObserver:self forKeyPath:@"isFinished" context:NULL];
+    [_dependencies removeObject:operation];
```

Comment:
On the reference platform, nothing happens. I'll add a test.

---
In reply to: [69623943](https://github.com/Microsoft/WinObjC/pull/602#discussion_r69623943) [](ancestors = 69623943)